### PR TITLE
Refactor DSG - changes for DB plugins

### PR DESCRIPTION
**Pull request by abrl91**
Issue Number: part of #3769 and #3768 

## PR Details

implement changes for DB plugins

## PR Checklist
- [ ] Tests for the changes have been added
- [x] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

### CODE REVIEW

File: packages/amplication-data-service-generator/src/server/prisma/create-prisma-schema.ts [hunk 1]
```diff
@@ -55,7 +55,13 @@ export async function createPrismaSchemaInternal({
     return enumFields.map((field) => createPrismaEnum(field, entity));
   });
 
-  const schema = PrismaSchemaDSL.createSchema(models, enums, dataSource, [
+  const prismaDataSource = {
+    name: PrismaSchemaDSL.DataSourceProvider[dataSource.provider].toLowerCase(),
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/prisma/create-prisma-schema.ts:59]:
```suggestion
    name: dataSource.name,
```

File: packages/amplication-data-service-generator/src/server/prisma/create-prisma-schema.ts [hunk 2]
```diff
@@ -55,7 +55,13 @@ export async function createPrismaSchemaInternal({
...
+    provider: PrismaSchemaDSL.DataSourceProvider[dataSource.provider],
+    url: new PrismaSchemaDSL.DataSourceURLEnv(dataSource.urlEnv),
+  };
+
+  const schema = PrismaSchemaDSL.createSchema(models, enums, prismaDataSource, [
     clientGenerator,
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/prisma/create-prisma-schema.ts:65]: To completely remove the dependency on prisma-schem-dsl from the plugins, we should also replace this object with an internal type 
```suggestion
    clientGenerator,
```

File: packages/amplication-data-service-generator/src/util/prepare-yaml-file.ts [hunk 3]
```diff
@@ -1,11 +1,16 @@
...
-  updateProperties: { [key: string]: any }
+  updateProperties: { path: string; value: any }[]
 ): string {
   const parsed = YAML.parse(yamlFileContent);
-  const updated = Object.assign(parsed, updateProperties);
+  if (!updateProperties.length) return YAML.stringify(parsed, { nullStr: "~" });
+
+  const updated = updateProperties.map((item) =>
+    set(parsed, item.path, item.value)
+  )[0];
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/util/prepare-yaml-file.ts:13]: although set() returns the mutated file - using map() will return an array of files 

I see by the results of the snapshot that somehow it is working - but I still think we should update this code 

you can use forEach or other ways to iterate and then just use the mutated "parsed" object

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]
Looks good. see my comments below.

File: packages/amplication-data-service-generator/src/server/prisma/constants.ts [hunk 4]
```diff
@@ -1,13 +1,13 @@
-import { PrismaDataSource } from "@amplication/code-gen-types";
+import { PrismaDataSource, PrismaClientGenerator } from "@amplication/code-gen-types";
 import * as PrismaSchemaDSL from "prisma-schema-dsl";
 
-export const CLIENT_GENERATOR: PrismaSchemaDSL.Generator = PrismaSchemaDSL.createGenerator(
+export const CLIENT_GENERATOR: PrismaClientGenerator = PrismaSchemaDSL.createGenerator(
   "client",
   "prisma-client-js"
 );
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/prisma/constants.ts:10]: should we sue the new type here?

File: packages/amplication-data-service-generator/src/server/prisma/constants.ts [hunk 5]
```diff
@@ -1,13 +1,13 @@
-import { PrismaDataSource } from "@amplication/code-gen-types";
+import { PrismaDataSource, PrismaClientGenerator } from "@amplication/code-gen-types";
 import * as PrismaSchemaDSL from "prisma-schema-dsl";
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/prisma/constants.ts:5]: import should be removed

File: packages/amplication-data-service-generator/src/server/prisma/create-prisma-schema.ts [hunk 6]
```diff
@@ -55,7 +55,13 @@ export async function createPrismaSchemaInternal({
     return enumFields.map((field) => createPrismaEnum(field, entity));
   });
 
-  const schema = PrismaSchemaDSL.createSchema(models, enums, dataSource, [
+  const prismaDataSource = {
+    name: dataSource.name,
+    provider: PrismaSchemaDSL.DataSourceProvider[dataSource.provider],
+    url: new PrismaSchemaDSL.DataSourceURLEnv(dataSource.urlEnv),
+  };
+
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/prisma/create-prisma-schema.ts:63]: you need to create the clientGenerator here based on the input

File: packages/amplication-data-service-generator/src/tests/__snapshots__/create-data-service-custom-path.spec.ts.snap [hunk 7]
```diff
@@ -2867,7 +2867,9 @@ export class StringNullableFilter {
 ",
   "test/.env": "POSTGRESQL_USER=testUsername
 POSTGRESQL_PASSWORD=1234
+POSTGRESQL_DB_NAME=/db-name
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/tests/__snapshots__/create-data-service-custom-path.spec.ts.snap:2870]: is that really supposed to include /

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

File: packages/amplication-data-service-generator/src/server/prisma/constants.ts (hunk 5 above)
Comment (abrl91) [reply packages/amplication-data-service-generator/src/server/prisma/constants.ts:5]: which import? I need all of them
Screen Shot 2022-09-17 at 13 50 48

File: packages/amplication-data-service-generator/src/server/prisma/create-prisma-schema.ts [hunk 8]
```diff
@@ -55,7 +55,13 @@ export async function createPrismaSchemaInternal({
...
+  const schema = PrismaSchemaDSL.createSchema(models, enums, prismaDataSource, [
     clientGenerator,
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/prisma/create-prisma-schema.ts:65]:
```suggestion
    PrismaSchemaDSL.createGenerator(
  clientGenerator.name,
clientGenerator.provider
);,
```

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

**Review by yuval-hazaz — APPROVED** [review]

**Pull request merged**